File: Stub/JungleGymStub/LiveViewManager.m
//
//  LiveViewManager.m
//  JungleGymStub
//
//  Created by Brandon Evans on 2017-11-19.
//  Copyright © 2017 Brandon Evans. All rights reserved.
//

#import "LiveViewManager.h"
@import Foundation;
@import UIKit;

@implementation LiveViewManager

- (void)setWindow:(UIWindow *)window {
    if ([window isEqual:_window]) { return; }

    _window = window;

    if (window) {
        [window makeKeyAndVisible];
        [self sendLiveViewAvailable];
    }
    else {
        [self sendLiveViewDismissed];
    }
}

- (void)setViewController:(UIViewController *)viewController {
    if ([viewController isEqual:_viewController]) { return; }

    _viewController = viewController;

    if (viewController) {
        self.view = nil;
        self.window = [self windowForPresentingLiveViewController:viewController];
    }
    else {
        self.window = nil;
    }
}

- (void)setView:(UIView *)view {
    if ([view isEqual:_view]) { return; }

    _view = view;

    if (view) {
        self.viewController = nil;
        self.window = [self windowForPresentingLiveView:view];
    }
    else {
        self.window = nil;
    }
}

- (void)sendLiveViewAvailable {
}

- (void)sendLiveViewDismissed {
    NSDictionary *properties = @{@"date": [NSDate date], @"dismissed": @YES};
    NSData *propertiesData = [NSPropertyListSerialization dataWithPropertyList:properties format:NSPropertyListBinaryFormat_v1_0 options:0 error:nil];
    // DVTPlaygroundCommunicationSender.shared.send(data: data, identifier: "XCPLiveView", version: "XCPLiveView", completion: { error in })
}

- (UIWindow *)windowForPresentingLiveViewController:(UIViewController *)viewController {
    UIWindow *window = [self windowForHostingLiveView];
    CGSize size = [viewController preferredContentSize];
    window.frame = CGRectMake(0, 0, size.width, size.height);
    window.rootViewController = viewController;
    return window;
}

- (UIWindow *) windowForPresentingLiveView:(UIView *)view {
    UIWindow *window = [self windowForHostingLiveView];
    window.frame = CGRectMake(0, 0, view.bounds.size.width, view.bounds.size.height);
    [window addSubview:view];
    return window;
}

- (UIWindow *)windowForHostingLiveView {
    UIWindow *window = [[UIWindow alloc] init];
    window.screen = [UIScreen mainScreen];
    window.backgroundColor = [UIColor blackColor];
    window.opaque = YES;
    return window;
}

@end

File: Stub/JungleGymStub/LiveViewManager.swift
//
//  LiveViewManager.swift
//  JungleGymStub
//
//  Created by Brandon Evans on 2017-11-16.
//  Copyright © 2017 Brandon Evans. All rights reserved.
//

import Foundation
import UIKit

class LiveViewManager {
    var window: UIWindow? {
        didSet {
            guard window != oldValue else { return }

            if let window = window {
                window.makeKeyAndVisible()
                sendLiveViewAvailable()
            }
            else {
                sendLiveViewDismissed()
            }
        }
    }

    var viewController: UIViewController? {
        didSet {
            guard viewController != oldValue else { return }

            if let viewController = viewController {
                view = nil
                window = windowForPresentingLiveViewController(viewController)
            }
            else {
                window = nil
            }
        }
    }

    var view: UIView? {
        didSet {
            guard view != oldValue else { return }

            if let view = view {
                viewController = nil
                window = windowForPresentingLiveView(view)
            }
            else {
                window = nil
            }
        }
    }

    func sendLiveViewDismissed() {
        let properties: [String : Any] = ["date": Date(), "dismissed": true]
        /*let data*/ _ = try? PropertyListSerialization.data(fromPropertyList: properties, format: PropertyListSerialization.PropertyListFormat.binary, options: PropertyListSerialization.WriteOptions(bitPattern: 0))
        // DVTPlaygroundCommunicationSender.shared.send(data: data, identifier: "XCPLiveView", version: "XCPLiveView", completion: { error in })
    }

    func sendLiveViewAvailable() {
        guard
            let view = view ?? viewController?.view,
            let window = window
        else { return }

        let properties: [String : Any] = [
            "date": Date(),
            "width": view.bounds.width,
            "height": view.bounds.height,
            "windowX": window.frame.origin.x,
            "windowY": window.frame.origin.y,
            "screenScale": window.screen.scale
        ]
        /*let data*/ _ = try? PropertyListSerialization.data(fromPropertyList: properties, format: PropertyListSerialization.PropertyListFormat.binary, options: PropertyListSerialization.WriteOptions(bitPattern: 0))
        // DVTPlaygroundCommunicationSender.shared.send(data: data, identifier: "XCPLiveView", version: "XCPLiveView", completion: { error in })
    }

    func windowForPresentingLiveViewController(_ viewController: UIViewController) -> UIWindow {
        let window = windowForHostingLiveView()
        let size = viewController.preferredContentSize
        window.frame = CGRect(x: 0, y: 0, width: size.width, height: size.height)
        window.rootViewController = viewController
        return window
    }

    func windowForPresentingLiveView(_ view: UIView) -> UIWindow {
        let window = windowForHostingLiveView()
        window.frame = CGRect(x: 0, y: 0, width: view.bounds.width, height: view.bounds.height)
        window.addSubview(view)
        return window
    }

    func windowForHostingLiveView() -> UIWindow {
        let window = UIWindow()
        window.screen = UIScreen.main
        window.backgroundColor = .black
        window.isOpaque = true
        return window
    }
}
